### ISSUE TITLE
[bug] cursor out of window error

### ISSUE THREAD

**Issue by ViRu-ThE-ViRuS** [Wansmer/treesj#11]
```
|| Error executing Lua callback: ...e/nvim/site/pack/packer/opt/treesj/lua/treesj/format.lua:67: Cursor position outside buffer
|| stack traceback:
|| 	[C]: in function 'nvim_win_set_cursor'
|| 	...e/nvim/site/pack/packer/opt/treesj/lua/treesj/format.lua:67: in function '_format'
|| 	...are/nvim/site/pack/packer/opt/treesj/lua/treesj/init.lua:20: in function <...are/nvim/site/pack/packer/opt/treesj/lua/treesj/init.lua:19>
```

I am woking on adding support for C++, and encountered the above trace with the config:

``` lua
treesj.setup({
    langs = {
        cpp = {
            compound_statement = treesj_utils.set_preset_for_list({
                both = {
                    separator = ';',
                    last_separator = true,
                    no_format_with = {}
                }
            })
        },
    },
})
```

in the code segment:

``` cpp
  if (true) {
     ^^^^^ cursor anywhere before `{`
    int a = 10;
    int b = 15;
  }
```

using the `TSJSplit`. The entire document is messed up after this crash as well. 
`TSJJoin` in the same case, gives the error: `Cannot join node longer ....`. 

otherwise, it works as expected:
```cpp
  if (true) {
              ^ cursor at or after `{`
    int a = 10;
    int b = 15;
  }

---

  if (true) { int a = 10; int b = 15; }
```

For reference, the entire block `{ .... }` is the TS node `compound_statement`.

NOTE:
I know this use case does not make sense because why would you put different statements on the same line, but its just something I encountered while playing around. This is a bug though, and might be encountered elsewhere, because this config is 'correct'.

**Comment by Wansmer** [comment]
I can't reproduce this bug. Maybe you're calling plugin when this fragment inside another node and the error happen on it?
Nevertheless, I will look for the error. Thanks for helpings.

https://user-images.githubusercontent.com/46977173/203769307-648bde07-87fc-4296-a44b-aadbd24455ec.mov

**Comment by ViRu-ThE-ViRuS** [comment]
Hey! I cannot reproduce the bug either anymore, it might be I was on an older version of plugin or something. But you have given me a workable idea:

Is it possible to only affect the `compound_statements` inside an `if_statement`, not in general?
Because this config (as can be seen in your comment), will change the entire function if im outside an `if_statement` because it is a `compound_statement`.

**Comment by Wansmer** [comment]
I found a case, why error presented: 
( code from [luajit-remake](https://github.com/luajit-remake/luajit-remake) )
```cpp
static void NO_RETURN HandleMetatableSlowPath(TValue tvIndex, TValue base, TValue metamethod) {
  while (true) {
    // Calling plugin ':TSJSplit' when cursor at 'likely' { row 9, col 13 }.
    // The 'if_statement' NOT configured with "{ target_nodes = { 'compound_statement' } }". 
    // TreeSJ continue search for suitable node up and found 'compound_statement' at line(row) 2. 
    // 'compound_statement' is already splitted. Repeated split (without "split = { recursive = true }")
    // is joining all nested multiline node to one line.
    if (likely(metamethod.Is<tHeapEntity>())) {
      HeapEntityType mmType = metamethod.GetHeapEntityType();
      if (mmType == HeapEntityType::Function) {
        MakeCall(metamethod.As<tFunction>(), base, tvIndex,
                 GlobalGetMetamethodCallContinuation);
      } else if (mmType == HeapEntityType::Table) {
        assert(tvIndex.Is<tString>());
        HeapPtr<HeapString> index = tvIndex.As<tString>();
        HeapPtr<TableObject> tableObj = metamethod.As<tTable>();
        GetByIdICInfo icInfo;
        TableObject::PrepareGetById(
            tableObj, UserHeapPointer<HeapString>{index}, icInfo /*out*/);
        TValue result = TableObject::GetById(tableObj, index, icInfo);
        if (likely(!icInfo.m_mayHaveMetatable || !result.Is<tNil>())) {
          Return(result);
        }
        EnterSlowPath<CheckMetatableSlowPath>(tableObj);
      }
    }
  }
}
```
When trying 'split' already split block, sticky-cursor work wrong. I'll think, how correctly handle this. At the moment, I just hide error with safe calling.

**Comment by Wansmer** [comment]
> Is it possible to only affect the `compound_statements` inside an `if_statement`, not in general?

Not sure what I correctly understood your question, but '{ target_nodes = { ... } }' just for it.

> Some nodes can be seen to have `names` as well in TS Playground, can these names/titles be supported in TreesJ, not just the name of the node?

Hmm, you mean like it configured in [javascript](https://github.com/Wansmer/treesj/blob/992683e381682397d9b5c22d06adcdbba0c919d0/lua/treesj/langs/javascript.lua#L45) for `try_statement` what contains few `statement_block` in `catch_clause` and `finaly_clause`?

**Comment by ViRu-ThE-ViRuS** [comment]
> > Some nodes can be seen to have `names` as well in TS Playground, can these names/titles be supported in TreesJ, not just the name of the node?
> 
> Hmm, you mean like it configured in [javascript](https://github.com/Wansmer/treesj/blob/992683e381682397d9b5c22d06adcdbba0c919d0/lua/treesj/langs/javascript.lua#L45) for `try_statement` what contains few `statement_block` in `catch_clause` and `finaly_clause`?

<img width="736" alt="Screen Shot 2022-11-25 at 22 29 28" src="https://user-images.githubusercontent.com/16427330/204075521-7d6e219b-ddc6-4ea5-b7d3-2d8387e1919e.png">

Like in this setup, can the red highlighted `condition` and `consequence` and `alternative` be added as nodes to the config? EG:

``` lua
javascript = {

-- is this possible?
    condition = { --[[ ]] },
    consequence = { --[[ ]] },
    alternative = { --[[ ]] },

-- instead of this
    statement_block = { -- [[ ]] },
    parenthesized_expression = { --[[ ]] },
    else_clause = { -- [[ ]] }
}
```

this way, only say `condition` node for `if_statement` will be effect, not any other `parenthesized_expression` nodes in the program. 

> > Is it possible to only affect the `compound_statements` inside an `if_statement`, not in general?
> 
> Not sure what I correctly understood your question, but '{ target_nodes = { ... } }' just for it.

this is related to the above as well.
Basically, the problem im facing is:

https://user-images.githubusercontent.com/16427330/204075925-a5ab1f26-6692-41bc-b6d1-e5adc271b51d.mov

IE: with the config:
```lua
cpp = {
            compound_statement = treesj_utils.set_preset_for_list({
                both = {
                    separator = ';',
                    last_separator = true,
                    no_format_with = {}
                }
            }),
            if_statement = { target_nodes = { 'compound_statement' }}
}
```
`Split/Join` work perfectly when on/inside the `if_statement`. 
But `Split` when outside the `if_statement`, but still inside the function body (which itself is a `compound_statement`), will destroy the formatting of the function body, and also place the cursor at a seemingly weird position. `Join` in this case makes the function body in one line.

I hope to achieve the behavior where, `if_statement` still works the same as above, 
but calling `Split/Join` has no effect when outside `if_statement`. 

This can be made possible, as I mention in the previous reply:

<img width="710" alt="Screen Shot 2022-11-25 at 22 48 35" src="https://user-images.githubusercontent.com/16427330/204076198-4280abb7-c2cb-42f6-8384-fdbf1145ac9a.png">

If we can just add configs for `consequence` which is specific, but no need for `compound_statement` which is general.

**Comment by Wansmer** [comment]
It cannot detect 'if_statement' outside 'if_statement'. And I don't understand for why it can be need? :-) Here, plugin found a `compound_statement` of function because it is the closest suitable node. See ['How it works'](https://github.com/Wansmer/treesj#how-it-works) in readme.

You can strongly restrict split/join 'compound_statement' if it contains 'if_statement':
```lua
local cpp = {
  compound_statement = u.set_preset_for_list({
    both = {
      separator = ';',
      last_separator = true,
      no_format_with = { 'if_statement' },
    },
  }),
  if_statement = { target_nodes = { 'compound_statement' } },
}
```
In my plan, implement functionality, when it will be able to configure same node the different depending on the context. Maybe it will solve it task.

**Comment by ViRu-ThE-ViRuS** [comment]
> It cannot detect 'if_statement' outside 'if_statement'. And I don't understand for why it can be need? :-) Here, plugin found a `compound_statement` of function because it is the closest suitable node. See ['How it works'](https://github.com/Wansmer/treesj#how-it-works) in readme.
> 
> You can strongly restrict split/join 'compound_statement' if it contains 'if_statement':
> 
> ```lua
> local cpp = {
>   compound_statement = u.set_preset_for_list({
>     both = {
>       separator = ';',
>       last_separator = true,
>       no_format_with = { 'if_statement' },
>     },
>   }),
>   if_statement = { target_nodes = { 'compound_statement' } },
> }
> ```
> 

I see, should have studied the README.md more :'). Thank you!
Although, Ive seen to be a better/more general config:
``` lua
local cpp = {
  compound_statement = u.set_preset_for_statement({
    both = {
      separator = ';',
      last_separator = true,
      no_format_with = { 'compound_statement' },
    },
  }),
  if_statement = { target_nodes = { 'compound_statement' } },
}
```

> In my plan, implement functionality, when it will be able to configure same node the different depending on the context. Maybe it will solve it task.

YES! this would definitely give more control to the user as well as the plugin itself! I vote its a good feature to add.
Ill go update the overall config ive found to be generally useful for Cpp / Python in PRs!

**Comment by ViRu-ThE-ViRuS** [comment]
Keeping issue open till original error is resolved.

**Comment by Wansmer** [comment]
> YES! this would definitely give more control to the user as well as the plugin itself! I vote its a good feature to add. Ill go update the overall config ive found to be generally useful for Cpp / Python in PRs!

Excellent! Please, add tests in a PR according to ['tests/README.md'](https://github.com/Wansmer/treesj/blob/main/tests/README.md)

**Comment by Wansmer** [comment]
Fixed in #18
